crates/lunatic-sqlite-api/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+
+    // get the query
+    let query = memory_slice
+        .get(query_str_ptr as usize..(query_str_ptr + query_str_len) as usize)
+        .or_trap("lunatic::sqlite::query_prepare::get_query")?;
+    let query = std::str::from_utf8(query).or_trap("lunatic::sqlite::query_prepare::from_utf8")?;
+
+    // obtain the sqlite connection
+    let conn = state
+        .sqlite_connections()
+        .get(conn_id)
+        .or_trap("lunatic::sqlite::query_prepare::obtain_conn")?
+        .lock()
+        .or_trap("lunatic::sqlite::query_prepare::obtain_conn")?;
+
+    // prepare the statement
+    let mut statement = conn
+        .prepare(query)
+        .or_trap("lunatic::sqlite::query_prepare::prepare_statement")?;
+
+    // allocate a vec to return some bytes back to the module
+    let mut return_value = Vec::new();
+
+    while let Ok(sqlite::State::Row) = statement.next() {
+        let count = statement.column_count();
+        for i in 0..count {
+            let mut bytes = match statement.column_type(i) {
+                sqlite::Type::Binary => {
+                    let mut bytes = statement
+                        .read::<Vec<u8>>(i)
+                        .or_trap("lunatic::sqlite::query_prepare::read_binary")?;
+
+                    let len = bytes.len();
+                    let mut result = vec![ColumnType::Binary as u8];
+                    result.append(&mut (len as u32).to_le_bytes().to_vec());
+                    result.append(&mut bytes);
+                    result
+                }
+                sqlite::Type::Float => {
+                    let mut result = vec![ColumnType::Float as u8];
+                    result.append(
+                        &mut statement
+                            .read::<f64>(i)
+                            .or_trap("lunatic::sqlite::query_prepare::read_float")?
+                            .to_le_bytes()
+                            .to_vec(),
+                    );
+                    result
+                }
+                sqlite::Type::Integer => {
+                    let mut result = vec![ColumnType::Integer as u8];
+                    result.append(
+                        &mut statement
+                            .read::<i64>(i)
+                            .or_trap("lunatic::sqlite::query_prepare::read_integer")?
+                            .to_le_bytes()
+                            .to_vec(),
+                    );
+                    result
+                }
+                sqlite::Type::String => {
+                    let bytes = statement
+                        .read::<String>(i)
+                        .or_trap("lunatic::sqlite::query_prepare::read_string")?;
+
+                    let len = bytes.len();
+                    let mut result = vec![ColumnType::String as u8];
+                    result.append(&mut (len as u32).to_le_bytes().to_vec());
+                    result.append(&mut bytes.as_bytes().to_vec());
+                    result
```

Comment:
Instead of appending to a temporary vector (`result`), we could here directly append to `return_value`.